rfd/0197-better-async-state-management-with-tanstack-query.md

```diff
@@ -0,0 +1,266 @@
+- Paginated queries
+  - At the moment we have a large wrapper utility to add paginated query support for our tables. TanStack Query provides a built-in
+    way to handle paginated queries, which is much easier to use and understand
+- Parallel queries
+  - TanStack Query can allow us to run multiple queries in parallel, which is useful for loading multiple pieces of data at once
+- Suspense support
+  - As it stands we're not using React's Suspense API, but TanStack Query has built-in support for it. This can allow us to use Suspense for data fetching, which can simplify our code and improve the user experience
+
+### Suspense support
+
+Suspense is a great way to fetch data, moving the loading and error states into the component tree rather than
+every component that fetches data dealing with the loading and error states.
+
+For example, instead of
+
+```tsx
+// psuedo code
+
+const attempt = useAttempt();
+
+if (attempt.status === 'processing') {
+  return <Loading />;
+}
+
+if (attempt.status === 'failed') {
+  return <Error />;
+}
+
+// continue rendering the component
+```
+
+The component can be wrapped in a `Suspense` component, which will show a loading state until the data is ready:
+
+```tsx
+<ErrorBoundary FallbackComponent={BotInformationError}>
+  <Suspense fallback={<BotInformationSkeleton />}>
+    <BotInformationInner {...props} />
+  </Suspense>
+</ErrorBoundary>
+```
+
+This keeps the logic of `BotInformationInner` tightly coupled to the data it needs, rather than having to deal with
+loading and error states - they're handled by the component tree.
+
+There are other benefits to using Suspense - it gives more control over the loading state when there are multiple
+components that need to fetch data, as well as allowing for more granular error handling. [Read more here](https://react.dev/reference/react/Suspense).
+
+### Example query
+
+```tsx
+const { data, error, isPending } = useQuery({
+  queryKey: ['accessList', accessListId],
+  queryFn: () => accessManagementService.fetchAccessList(accessListId),
+});
+```
+
+Which could be wrapped as a hook for easy reuse:
+
+```tsx
+export function useGetAccessList(id: string) {
+  return useQuery({
+    queryKey: ['accessList', id],
+    queryFn: () => accessManagementService.fetchAccessList(id),
+  });
+}
+```
+
+### Example mutation
+
+```tsx
+const { mutateAsync: deleteAccessList, isPending } = useMutation({
+  mutationFn: (accessListId: string) =>
+    accessManagementService.deleteAccessList(accessListId),
+});
+
+async function onOk() {
+  await deleteAccessList(accessListId);
+
+  history.push(cfg.getAccessListManagementRoute(), {
+    deletedAccessListId: accessListId,
+  });
+}
+```
+
+### Example infinite query
+
+```tsx
+const { fetchNextPage, hasNextPage, data, isPending } = useInfiniteQuery({
+  getNextPageParam: lastPage => lastPage.next_cursor,
+  initialPageParam: '',
+  queryFn,
+  queryKey,
+});
+```
```

Comment:
i havent looked up infinite queries, but does it expect a specific return type? or we can define the keys for each of this based on the return value from the backend?

Reply:
No specific return type, it works with any backend response